Nodejs/Product/Nodejs/Workspace/PackageJsonTestContainerDiscoverer.cs

```diff
@@ -0,0 +1,147 @@
+                }
+            }
+        }
+
+        public event EventHandler TestContainersUpdated;
+
+        private async Task AttemptUpdateAsync()
+        {
+            if (this.currentWorkspace != null)
+            {
+                var indexServce = this.currentWorkspace.GetIndexWorkspaceService();
+                var filesDataValues = await indexServce.GetFilesDataValuesAsync<string>(NodejsConstants.TestRootDataValueGuid);
+
+                lock (this.containerLock)
+                {
+                    this.containers.Clear();
+                    foreach (var dataValue in filesDataValues)
+                    {
+                        var rootFilePath = this.currentWorkspace.MakeRooted(dataValue.Key);
+                        var testRoot = dataValue.Value.Where(f => f.Name == "TestRoot").Select(f => f.Value).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(testRoot))
+                        {
+                            var testRootPath = this.currentWorkspace.MakeRooted(testRoot);
+                            this.containers.Add(new PackageJsonTestContainer(this, rootFilePath, testRootPath));
+                        }
+                    }
+                }
+            }
+            this.TestContainersUpdated?.Invoke(this, EventArgs.Empty);
+        }
+
+        private async Task OnActiveWorkspaceChangedAsync(object sender, EventArgs e)
+        {
+            this.currentWorkspace = this.workspaceService.CurrentWorkspace;
+
+            await AttemptUpdateAsync();
+        }
+
+        private Task FileSystemChangedAsync(object sender, FileSystemEventArgs args)
+        {
+            // We only need to raise the containers updated event in case a js file contained in the
+            // test root is updated.
+            // Any changes to the 'package.json' will be handled by the FileScannerCompleted event.
+            if (IsJavaScriptFile(args.FullPath))
+            {
+                // use a flag so we don't deadlock
+                var testsUpdated = false;
+                lock (this.containerLock)
+                {
+                    foreach (var container in this.containers)
+                    {
+                        if (container.IsContained(args.FullPath))
+                        {
+                            container.IncreaseVersion();
+                            testsUpdated = true;
+                            break;
+                        }
+                    }
+                }
+                if (testsUpdated)
+                {
+                    this.TestContainersUpdated?.Invoke(this, EventArgs.Empty);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private async Task FileScannerCompletedAsync(object sender, FileScannerEventArgs args)
+        {
+            await AttemptUpdateAsync();
+        }
+
+        private static bool IsJavaScriptFile(string path)
```

Comment:
>IsJavaScriptFile [](start = 28, length = 16)

I'm skeptical that there isn't a helper for this or, at least, constants for known file extensions.

Reply:
There really isn't a helper for detecting JS files?

---
In reply to: [188796597](https://github.com/Microsoft/nodejstools/pull/1959#discussion_r188796597) [](ancestors = 188796597)

Reply:
Nope. 

I'll refactor that in another PR